### PR TITLE
More Windows installer improvements

### PR DESCRIPTION
**Pull request by arsing**
- `Uninstall-SecurityDaemon` now defaults to leaving config.yaml and
  Moby data root, unless new flags `-DeleteConfig` and `-DeleteMobyDataRoot`
  are specified.

- `Uninstall-SecurityDaemon` now deletes the edge containers instead of just
  stopping them, to ensure that they're deleted even if the Moby data root
  isn't.

- `Install-SecurityDaemon` now has a new flag `-ExistingConfig` that can be
  set to use an existing config.yaml, instead of generating a new one via
  `-Manual` or `-Dps`

- Fixed a bug in `Uninstall-SecurityDaemon` where it incorrectly parsed the
  result of `docker inspect`.

### CODE REVIEW

File: scripts/windows/setup/IotEdgeSecurityDaemon.ps1 [hunk 1]
```diff
@@ -361,7 +387,9 @@ function Get-SecurityDaemon {
                     -Delete ([ref] $deleteMobyEngineArchive)
             Expand-Archive $mobyEngineArchivePath $MobyInstallDirectory -Force
 
-            New-Item -Type Directory $MobyDataRootDirectory | Out-Null
+            if (-not (Test-Path $MobyDataRootDirectory)) {
```
Comment (damonbarry) [inline scripts/windows/setup/IotEdgeSecurityDaemon.ps1]: A simpler option is to use the `-Force` flag on the `New-Item` command. We're already using that approach when we create the domain socket parent folders.

File: scripts/windows/setup/IotEdgeSecurityDaemon.ps1 [hunk 2]
```diff
@@ -849,23 +904,23 @@ function Get-AgentRegistry {
...
-function Stop-IotEdgeContainers {
+function Remove-IotEdgeContainers {
     $dockerExe = Get-DockerExePath
 
     if (-not (Test-IsDockerRunning 6> $null)) {
         return
     }
 
-    $runningContainersString = Invoke-Native "$dockerExe ps --format ""{{.ID}}""" -Passthru
+    $runningContainersString = Invoke-Native "$dockerExe ps --all --format ""{{.ID}}""" -Passthru
```
Comment (damonbarry) [inline scripts/windows/setup/IotEdgeSecurityDaemon.ps1]: Thanks...I noticed this bug in the last couple days, but didn't know the cause...

File: scripts/windows/setup/IotEdgeSecurityDaemon.ps1 [hunk 3]
```diff
@@ -849,23 +904,23 @@ function Get-AgentRegistry {
...
 
     foreach ($containerId in $runningContainers) {
-        $inspectString = Invoke-Native "$dockerExe inspect ""$containerId""" -Passthru | ConvertFrom-Json
+        $inspectString = Invoke-Native "$dockerExe inspect ""$containerId""" -Passthru
         $inspectResult = ($inspectString | ConvertFrom-Json)[0]
 
         if ($inspectResult.Config.Labels.'net.azure-devices.edge.owner' -eq 'Microsoft.Azure.Devices.Edge.Agent') {
-            Invoke-Native "$dockerExe stop ""$containerId"""
-            Write-Verbose "Stopped container $($inspectResult.Name)"
+            Invoke-Native "$dockerExe rm --force ""$containerId"""
```
Comment (damonbarry) [inline scripts/windows/setup/IotEdgeSecurityDaemon.ps1]: I'm not sure if we want to delete the containers when the customer upgrades their iotedged?

File: scripts/windows/setup/IotEdgeSecurityDaemon.ps1 (hunk 2 above)
Comment (arsing) [reply scripts/windows/setup/IotEdgeSecurityDaemon.ps1]: The bug is the double `ConvertFrom-JSON` actually.

File: scripts/windows/setup/IotEdgeSecurityDaemon.ps1 (hunk 3 above)
Comment (arsing) [reply scripts/windows/setup/IotEdgeSecurityDaemon.ps1]: We intentionally want to, to ensure that the new version gets pulled.

File: scripts/windows/setup/IotEdgeSecurityDaemon.ps1 (hunk 1 above)
Comment (arsing) [reply scripts/windows/setup/IotEdgeSecurityDaemon.ps1]: Whoops. Forgot about that.

File: scripts/windows/setup/IotEdgeSecurityDaemon.ps1 (hunk 2 above)
Comment (damonbarry) [reply scripts/windows/setup/IotEdgeSecurityDaemon.ps1]: Ah, the change a couple lines down...

File: scripts/windows/setup/IotEdgeSecurityDaemon.ps1 (hunk 3 above)
Comment (arsing) [reply scripts/windows/setup/IotEdgeSecurityDaemon.ps1]: As discussed, the label does get applied to user modules too, so this isn't correct. And this still leaves the container image behind anyway.

File: scripts/windows/setup/IotEdgeSecurityDaemon.ps1 [hunk 4]
```diff
@@ -33,48 +33,89 @@ enum ContainerOs {
...
 
+        # The DPS registration ID.
         [Parameter(Mandatory = $true, ParameterSetName = 'DPS')]
         [String] $RegistrationId,
 
+        # The base OS of all the containers that will be run on this device via the security daemon.
+        #
+        # If set to Linux, a separate installation of Docker for Windows is expected.
+        #
+        # If set to Windows, the Moby Engine will be installed automatically.
```
Comment (damonbarry) [inline scripts/windows/setup/IotEdgeSecurityDaemon.ps1]: Let people know their existing Docker for Windows installation won't be impacted in this case.

**Pull request merged**